src/core.c/CompUnit/PrecompilationRepository.pm6

```diff
@@ -174,19 +175,20 @@ Need to re-check dependencies.")
 
             if $resolve {
                 my str $serialized-id = $dependency.serialize;
-                nqp::ifnull(
-                  nqp::atkey($resolved,$serialized-id),
-                  nqp::if(do {
-                        my $comp-unit := $REPO.resolve($dependency.spec);
-                        $!RMD("Old id: $dependency.id(), new id: {
-                            $comp-unit and $comp-unit.repo-id
-                        }")
-                          if $!RMD;
-
-                        $comp-unit and $comp-unit.repo-id eq $dependency.id
-                    },
-                    $loaded-lock.protect({ nqp::bindkey($resolved,$serialized-id, 1) }),
-                    (return False)));
+                $resolved-lock.protect: {
+                    nqp::ifnull(
+                      nqp::atkey($resolved,$serialized-id),
+                      nqp::bindkey($resolved,$serialized-id, do {
```

Comment:
I think this would be less convoluted when changing the `do {}` to `nqp::stmts()` (or a block) and moving the `nqp::bindkey` after the `return False`.
(Or I miss the point. :-P)

Reply:
Honestly I would be surprised if all the nqp stuff here was worth it. This is not code that runs tens of thousands of times.